src/Data/Expr.purs

```diff
@@ -4,259 +4,157 @@ import Prelude
-  | Print (o ~ String) (Expr i)
+instance encodeJsonExpr :: EncodeJson Expr where
+  encodeJson = encodeJson <<< toExprA
 
-data Expr o = Expr (∀ e. (∀ i. Expressible i => ExprF i o -> e) -> e)
+instance decodeJsonExpr :: DecodeJson Expr where
+  decodeJson json = do
+    x' <- decodeJson json
+    x' .: "op" >>= case _ of
+      "Val" -> x' .: "out" >>= case _ of
+        "Boolean" -> x' .: "param" >>= pure <<< boolean_
+        "Int" -> x' .: "param" >>= pure <<< int_
+        "String" -> x' .: "param" >>= pure <<< string_
+        out -> Left $ out <> " not implemented"
+      "If" -> x' .: "params" >>= case _ of
+        [x, y, z] -> pure (if_ x y z)
+        _ -> Left "Expected 3 params"
+      "Equal" -> x' .: "params" >>= case _ of
+        [x, y] -> pure (equal_ x y)
+        _ -> Left "Expected 2 params"
+      "Lookup" -> x' .: "param" >>= pure <<< lookup_
+      op -> Left $ op <> " invalid op"
 
 data ExprA
   = ValA ExprType
   | IfA ExprA ExprA ExprA String String
   | EqualA ExprA ExprA String String
   | PrintA ExprA String String
-
-mkExpr
-  :: ∀ i o
-   . Expressible i
-  => Expressible o
-  => ExprF i o
-  -> Expr o
-mkExpr e = Expr (_ $ e)
-
-val_ :: forall o. Expressible o => o -> Expr o
-val_ x = mkExpr (Val identity x)
-
-if_
-  :: forall o
-   . Expressible o
-  => Expr Boolean
-  -> Expr o
-  -> Expr o
-  -> Expr o
-if_ x y z = mkExpr (If identity x y z)
-
-equal_
-  :: forall o
-   . Expressible o
-  => Expr o
-  -> Expr o
-  -> Expr Boolean
-equal_ x y = mkExpr (Equal identity x y)
-
-print_
-  :: forall o
-   . Expressible o
-  => Expr o
-  -> Expr String
-print_ x = mkExpr (Print identity x)
-
-runExpr
-  :: ∀ o e
-   . (∀ i. Expressible i => ExprF i o -> e)
-  -> Expr o
-  -> e
-runExpr r (Expr f) = f r
-
-evalExpr' :: ∀ o. Expressible o => Expr o -> o
-evalExpr' = runExpr eval
-  where
-  eval :: ∀ i. Expressible i => ExprF i o -> o
-  eval (Val _ x) = x
-  eval (If p x y z) = if evalExpr' x then evalExpr' y else evalExpr' z
-  eval (Equal p x y) = coerceSymm p $ evalExpr' x == evalExpr' y
-  eval (Print p x) = coerceSymm p $ print $ evalExpr' x
-
-toExprA :: ∀ o. Expressible o => Expr o -> ExprA
-toExprA = runExpr go
+  | LookupA Key String
+
+instance encodeJsonExprA :: EncodeJson ExprA where
+  encodeJson =
+    case _ of
+      ValA x -> "param" := x ~> encodeHelper "Val" "Void" (reflectType x)
+      IfA x y z i o ->
+        "params" := [ x, y, z ] ~> encodeHelper "If" i o
+      EqualA x y i o ->
+        "params" := [ x, y ] ~> encodeHelper "Equal" i o
+      PrintA x i o ->
+        "params" := [ x ] ~> encodeHelper "Print" i o
+      LookupA x o -> "param" := x ~> encodeHelper "Lookup" "Void" o
+
+toExprA :: Expr -> ExprA
+toExprA = case _ of
+  Val x -> ValA x
+  If x y z -> IfA (toExprA x) (toExprA y) (toExprA z) (reflectIn x) (reflectOut y)
+  Equal x y -> EqualA (toExprA x) (toExprA y) (reflectIn x) (reflectOut x)
+  Print x -> PrintA (toExprA x) (reflectIn x) "String"
+  Lookup x -> LookupA x "???"
```

Comment:
What do we do here? Does `LookupA` have an output type? Is it `Unit`? Is it something else?

Reply:
I think this should be the same type of whatever value you're looking up. Which should be known at encode time, since each input type has only one valid type. Also, there are type constraints on where in an expression it's used. 

Some clarifying examples:
```purescript
{ name: string_ "Name"
, visibility: lookup_ "active"
, description: if_ (lookup_ "active") (string_ "you need a name") (lookup_ "foo")
}
```
So above,  `visibility` needs to be a Boolean, so only boolean lookups are valid here, so this could only have been encoded if `active` is a Boolean input. Similarly, in `description`, the `active` lookup is valid because an `if_` condition has to be a Boolean expression.

Reply:
Hmm, it seems like we'd have to have the `Page Expr` then to do that, right? Or at least some `Map Key String`/`Key -> String` or whatever. The `Key`s only exist in a `Page Expr`, so maybe we can't have an instance for `EncodeJson Expr`. It would have to be `EncodeJson (Page Expr)` and friends, so they could traverse with the type context.

That doesn't seem too bad, since `ExprA`s reason for existence seems to be to facilitate encoding.

Reply:
I see what you mean, with `ExprF i o` we never lose our output type but if we drop our types then we have no way of recovering that information without additional context.

One trick we could try would be to have a failure case for our lookup, something like
```purescript
  | Lookup Key Expr
```
That way we can kind of cheat and look at the output type of the failure case Expr.

Reply:
That seems like it'll also remove one more error case to deal with.

Reply:
That seemed to work out! Great suggestion!